### DISCUSSION TITLE
Issue with Sentry Error Handling in Express.js

### DISCUSSION THREAD

**Discussion by yuuri111** [RobinTail/express-zod-api#1876] [Q&A] [answered]
Hello,

I would like to ask about the implementation of Sentry and error handling.

As you might already know, in Express.js, Sentry can be integrated as follows:
https://docs.sentry.io/platforms/javascript/guides/express/

I have configured Sentry in the file where createConfig() is called, as shown below. 
However, it seems that errors thrown in the endpoint (e.g., throw new Error()) are not being handled by the onError middleware.

```
Sentry.init({
  dsn: ***
});
Sentry.setupExpressErrorHandler(app);

app.use(function onError(err, req, res, next) {
  Sentry.captureException(err);
  res.status(500).send({ error: "error" });
});
```

Additionally, even when I set up an error handler as follows, the errors are not being handled properly:

```
const errorHandler = createResultHandler({
  getPositiveResponse: () => ({
    schema: z.string(),
    mimeType: "application/json",
  }),
  getNegativeResponse: () => ({
    schema: z.string(),
    mimeType: "application/json",
  }),
  handler: ({ response, error, output }) => {
    response.status(500).send("system error");
  },
});

export const config = createConfig({
  errorHandler,
});

```

What is the correct way to set up error handling in Express.js in this context?

Thank you for your assistance.

**Comment by RobinTail** [comment, marked as answer]
Hello @yuuri111 ,

Are you using `createServer()` method?

If so, the `errorHandler` you put into `createConfig()` is only used for the routing errors (when a route was not found). But the errors you `throw` from within your endpoints are handled by the corresponding ResultHandler of the EndpointsFactory that those endpoints were made on. 

Also, the `app.use` you call, is called after the proprietary error handler is installed, so it does not make any effect.

You need to make a custom result handler, similar to what you already made using `createResultHandler`, but more comprehensive:
- it should handle error by calling `Sentry.captureException`;
- it should handle no error by sending the response.

Assuming that you're using v19 or earlier, consider making a copy of the `defaultResultHandler`:
https://github.com/RobinTail/express-zod-api/blob/7d865da3487cfa9ad0045bcb07fb5d8ec1c74661/src/result-handler.ts#L69

But add a `Sentry.captureException(error)` to it, near the logging:

https://github.com/RobinTail/express-zod-api/blob/7d865da3487cfa9ad0045bcb07fb5d8ec1c74661/src/result-handler.ts#L109

Then do the following:

```ts
const factory = new EndpointsFactory(customResultHandler);
```

and `.build()` your endpoints from that factory.

I hope it helps.

**Reply by RobinTail** [reply]
Alternatively, you can use `attachRouting()` method instead. Then you can go with the approach advised by `Sentry` and do `app.use()` next, @yuuri111

**Comment by yuuri111** [comment]
@RobinTail 

Thank you for your confirmation and response.

>Are you using createServer() method?

Yes, I'm using createServer().
```
(async () => {
  await createServer(config, routing);
})();
```

Ideally, I would like to manage errors in a centralized manner, catching them all in one place, rather than setting Sentry.captureException in each handler.

Therefore, I would like to try using attachRouting(). Currently, I am using ServerConfig instead of AppConfig as shown below (because I need to use beforeRouting, etc.).

```
export const config = createConfig({
  server: {
    listen: 3000,
    beforeRouting: ({ app, logger }) => {
```

In this case (using ServerConfig), how can I use attachRouting?

Thank you for your assistance.

**Comment by RobinTail** [comment]
> rather than setting Sentry.captureException in each handler.

The original idea implies having one ResultHandler for consistent responses, but if they are not supposed to be consistent and  you're having multiple ResultHandlers then yes, it's probably better to go with `attachRouting()`.

> how can I use attachRouting?

Here is an article about that, @yuuri111 : 

https://github.com/RobinTail/express-zod-api/blob/master/README.md#connect-to-your-own-express-app

**Reply by yuuri111** [reply]
@RobinTail 

Thank you very much.
I checked the documentation, but I'm encountering the following error:

```
Argument of type 'ServerConfig<string>' is not assignable to parameter of type 'AppConfig<string>'.
Property 'app' is missing in type 'ServerConfig<string>' but required in type 'AppConfig<string>'.
```

**Reply by RobinTail** [reply]
You need to remove the `server` (and `https`) entry in the argument of `createConfig()` and replace it with `app`, @yuuri111

**Reply by yuuri111** [reply]
ok! Thank you so much!

**Comment by RobinTail** [comment]
Related discussion https://github.com/RobinTail/express-zod-api/discussions/1907
